python_modules/libraries/dagster-components/dagster_components/resolved/base.py

```diff
@@ -0,0 +1,271 @@
+        for field_name, resolver in field_resolvers.items()
+        # filter out _Unset to trigger defaults
+        if getattr(model, resolver.model_field_name or field_name) is not _Unset
+    }
+
+
+TType = TypeVar("TType", bound=type)
+
+
+def _safe_is_subclass(obj, cls: TType) -> TypeGuard[type[TType]]:
+    return (
+        isinstance(obj, type)
+        and not isinstance(obj, GenericAlias)  # prevent exceptions on 3.9
+        and issubclass(obj, cls)
+    )
+
+
+def _get_resolver(annotation: Any, field_name: str) -> "Resolver":
+    origin = get_origin(annotation)
+    args = get_args(annotation)
+
+    # explicit field level Resolver
+    if origin is Annotated:
+        resolver = next((arg for arg in args if isinstance(arg, Resolver)), None)
+        if resolver:
+            check.invariant(
+                _is_implicitly_resolved_type(args[0]) or resolver.model_field_type,
+                f"Resolver for {field_name} must define model_field_type {args[0]} is not model compliant.",
+            )
+            return resolver
+
+    if _is_implicitly_resolved_type(annotation):
+        return Resolver.default()
+
+    # nested or implicit
+    res = _dig_for_resolver(annotation, [])
+    if res:
+        return res
+
+    # legacy resolvers, to be removed
+    return derive_field_resolver(annotation, field_name)
+
+
+def _dig_for_resolver(annotation, path: Sequence[_TypeContainer]):
+    origin = get_origin(annotation)
+    args = get_args(annotation)
+    if _safe_is_subclass(annotation, Resolvable):
+        return Resolver(
+            partial(
+                _resolve_at_path,
+                container_path=path,
+                resolver=annotation.resolve_from_model,
+            ),
+            model_field_type=_wrap(annotation.model(), path),
+        )
+
+    if origin is Annotated:
+        resolver = next((arg for arg in args if isinstance(arg, Resolver)), None)
+        if resolver:
+            check.invariant(
+                _is_implicitly_resolved_type(args[0]) or resolver.model_field_type,
+                f"Nested resolver must define model_field_type {args[0]} is not model compliant.",
+            )
+            # need to ensure nested resolvers set their model type
+            return Resolver(
+                resolver.fn.__class__(
+                    partial(
+                        _resolve_at_path,
+                        container_path=path,
+                        resolver=resolver.fn.callable,
+                    )
+                ),
+                model_field_type=_wrap(resolver.model_field_type or args[0], path),
+            )
+
+    if origin in (Union, UnionType) and len(args) == 2:
+        left_t, right_t = args
+        if right_t is type(None):
+            res = _dig_for_resolver(left_t, [*path, _TypeContainer.OPTIONAL])
+            if res:
+                return res
+
+    elif origin in (Sequence, tuple, list):  # should look for tuple[T, ...] specifically
```

Comment:
should we error on `tuple[T, ...]`?

Reply:
`tuple[T, ...]` is the one that approximates to `Sequence[T]` that we want to hit on 

compared to `tuple[A, B, C]` fixed size tuple which we should handle different